Add Fabric support

Su5eD opened the pull request:
Adds a Fabric frontend DSL as well as a backend plugin to support generating fabric mod metadata.

### Frontend

The frontend DSL was created according to the [fabric.mod.json Specification](https://fabricmc.net/wiki/documentation:fabric_mod_json_spec) with support for all standard properties as well as custom fields. An example can be found inside the `TestFabric` subproject.

### Plugin

The backend plugin validates all field values where applicable. Currently, the only fields left out are `contact.irc` and `contact.sources` due to regex trouble. Fabric itself doesn't seem to actually validate these fields, so I'm not sure how strict we should be with URL values.

### Cross platform plugins

Added a new `ModsDotGroovyPlugin#getPlatforms` method that lets plugins provide a set of platforms they support. Plugins are then selected based on the enabled platform in gradle. This prevents them from interfering, at least until we have proper cross platform support in the future.

Su5eD committed:
Add Fabric DSL and Plugin

Initial Draft

Su5eD committed:
Load plugins depending on the configured platform(s)

Su5eD committed:
Set optional properties to null by default

Su5eD committed:
Fix nested contact verification

PaintNinja commented on TestFabric/src/main/resources/mods.groovy:
```diff
@@ -0,0 +1,124 @@
+        person "Maty"
+    }
+
+    custom {
+        it.'fabricMod:hello' = "world"
+    }
```
To avoid people trying `this.'fabricMod:hello' = "world"` and getting an unhelpful error, it might be a good idea to add a method called "property" or something.

PaintNinja commented on core/src/main/groovy/org/groovymc/modsdotgroovy/plugin/PluginRegistry.groovy:
```diff
@@ -15,9 +17,9 @@ final class PluginRegistry {
+    PluginRegistry(@Nullable Platform platform) {
         // Load plugins
-        plugins.addAll(ServiceLoader.load(ModsDotGroovyPlugin).asCollection())
+        plugins.addAll(ServiceLoader.load(ModsDotGroovyPlugin).findAll { ModsDotGroovyPlugin plugin -> platform == null || plugin.platforms.contains(platform) }.asCollection())
```
Please use `===` for null comparisons to avoid the overhead of going through Groovy Truth

PaintNinja commented on frontend-dsl-fabric/src/main/groovy/FabricModsDotGroovy.groovy:
```diff
@@ -0,0 +1,327 @@
+     */
+    @Nullable String icon = null
+
+    /**@
+     * Defines the licensing information.
+     * This should provide the complete set of preferred licenses conveying the entire mod package. In other words,
+     * compliance with all listed licenses should be sufficient for usage, redistribution, etc. of the mod package as a whole.
+     * For cases where a part of code is dual-licensed, choose the preferred license. The list is not exhaustive,
+     * serves primarily as a kind of hint, and does not prevent you from granting additional rights/licenses on a case-by-case basis.
+     * To aid automated tools, it is recommended to use <a href="https://spdx.org/licenses/">SPDX License Identifiers</a> for open-source licenses.
+     */
+    @Nullable def license = null
```
This all gets wrapped into one long paragraph. Might wanna put `<p></p>` for adding new lines.
<img width="557" alt="image" src="https://github.com/PaintNinja/ModsDotGroovy/assets/3158390/5f671827-40dc-408d-9da0-4103d10684cd">

```suggestion
    /**@
     * Defines the licensing information.
     * <p>This should provide the complete set of preferred licenses conveying the entire mod package. In other words,
     * compliance with all listed licenses should be sufficient for usage, redistribution, etc. of the mod package as a whole.</p>
     * <p>For cases where a part of code is dual-licensed, choose the preferred license. The list is not exhaustive,
     * serves primarily as a kind of hint, and does not prevent you from granting additional rights/licenses on a case-by-case basis.</p>
     * <p>To aid automated tools, it is recommended to use <a href="https://spdx.org/licenses/">SPDX License Identifiers</a> for open-source licenses.</p>
     */
    @Nullable def license = null
```

Likewise for other parts of the DSL.

PaintNinja commented on stock-plugins/src/main/groovy/org/groovymc/modsdotgroovy/plugin/FabricPlugin.groovy:
```diff
@@ -0,0 +1,269 @@
+
+                if (!value.containsKey("name")) {
+                    throw new PluginResult.MDGPluginException("Author name is required")
+                } else if (value.size() == 1) {
```
Please use `===` here

PaintNinja replied:
```suggestion
                } else if (value.size() === 1) {
```

PaintNinja commented on stock-plugins/src/main/groovy/org/groovymc/modsdotgroovy/plugin/FabricPlugin.groovy:
```diff
@@ -0,0 +1,269 @@
+                log.debug "contributors.contributor.onNestLeave: ${value}"
+                if (!value.containsKey("name")) {
+                    throw new PluginResult.MDGPluginException("Contributor name is required")
+                } else if (value.size() == 1) {
```
```suggestion
                } else if (value.size() === 1) {
```

PaintNinja commented on stock-plugins/src/main/groovy/org/groovymc/modsdotgroovy/plugin/FabricPlugin.groovy:
```diff
@@ -0,0 +1,269 @@
+
+        class Contributor {
+            String name
+            Contact contact = new Contact()
```
This is such a cool trick! :)

PaintNinja commented on frontend-dsl-fabric/src/main/groovy/org/groovymc/modsdotgroovy/frontend/fabric/PersonsBuilder.groovy:
```diff
@@ -0,0 +1,57 @@
+        log.debug "${fieldName}(name: $name)"
+        core.push(fieldName)
+        core.put('name', name)
+        if (closure != null) {
```
```suggestion
        if (closure !== null) {
```

PaintNinja commented on frontend-dsl-fabric/src/main/groovy/MixinsBuilder.groovy:
```diff
@@ -0,0 +1,44 @@
+        log.debug "mixins(config: $config)"
+        core.push('mixin')
+        core.put('config', config)
+        if (environment != null) {
```
```suggestion
        if (environment !== null) {
```

PaintNinja commented on TestFabric/src/main/resources/mods.groovy:
```diff
@@ -0,0 +1,124 @@
+        fabricloader = ">=0.14.21"
+        minecraft = "~1.20.1"
+        java = ">=17"
+        it.'fabric-api' = "*"
```
For better IDE support, a generic method should be added, e.g.:
```groovy
mod {
    modId = 'fabric-api'
    versionRange = '*'
}
mod 'fabric-api', '*'
```

PaintNinja commented on TestFabric/src/main/resources/mods.groovy:
```diff
@@ -0,0 +1,124 @@
+            value = "hello.kotlin.World"
+        }
+        client "net.fabricmc.example.ExampleModClient"
+        rei = "net.fabricmc.example.ReiPlugin"
```
Bug: setting rei twice adds rather than replaces:
```groovy
rei = "net.fabricmc.example.ReiPlugin"
rei = "net.fabricmc.new.ReiPlugin"
```
```json
"rei": [
    "net.fabricmc.example.ReiPlugin",
    "net.fabricmc.new.ReiPlugin"
],
```

PaintNinja reviewed (changes requested):
Some minor things but overall this looks great, thanks! :D

The large differences in fabric.mod.json vs mods.toml are interesting. I may have to reconsider some of mods.groovy v2's architecture and DSL convention to better suit this in future, but the idea of using methods for additive behaviour and handling lists manually in the plugin works well for now.

One thing to note is that the `ModsDotGroovyPlugin#getPlatforms` may get removed once proper multiplatform support is in the Gradle plugin, as there may be platform-agnostic plugins in the future (e.g. a spellchecker or HTTPS link enforcer).

Su5eD committed:
Apply suggestions from code review

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

Su5eD committed:
Apply requested changes

- Add property method to custom fields builder
- Add generic method to mod dependencies builder
- Fix entrypoint property setter overwrite
